File: app/src/main/java/fr/insapp/insapp/activities/MainActivity.java
package fr.insapp.insapp.activities;

import android.app.SearchManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.os.Build;
import android.os.Bundle;
import android.support.customtabs.CustomTabsClient;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.MenuItemCompat;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.preference.PreferenceManager;
import android.support.v7.widget.AppCompatCheckBox;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.CompoundButton;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.List;

import fr.insapp.insapp.App;
import fr.insapp.insapp.BuildConfig;
import fr.insapp.insapp.R;
import fr.insapp.insapp.adapters.ViewPagerAdapter;
import fr.insapp.insapp.fragments.ClubsFragment;
import fr.insapp.insapp.fragments.EventsFragment;
import fr.insapp.insapp.fragments.NotificationsFragment;
import fr.insapp.insapp.fragments.PostsFragment;
import fr.insapp.insapp.notifications.FirebaseMessaging;
import fr.insapp.insapp.utility.CustomTabsConnection;
import fr.insapp.insapp.utility.Utils;

public class MainActivity extends AppCompatActivity {

    public static final boolean dev = BuildConfig.DEBUG;

    public static CustomTabsConnection customTabsConnection;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // toolbar

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar_main);
        if (toolbar != null) {
            setSupportActionBar(toolbar);
            getSupportActionBar().setDisplayHomeAsUpEnabled(false);

            getSupportActionBar().setTitle(Utils.getUser().getUsername());
        }

        // view pager

        ViewPager viewPager = (ViewPager) findViewById(R.id.viewpager);
        setupViewPager(viewPager);

        // tab layout

        TabLayout tabLayout = (TabLayout) findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager);

        // custom tabs optimization

        MainActivity.customTabsConnection = new CustomTabsConnection();
        CustomTabsClient.bindCustomTabsService(this, "com.android.chrome", customTabsConnection);

        // Huawei protected apps

        ifHuaweiAlert();

        // topic notifications

        SharedPreferences defaultSharedPreferences = PreferenceManager.getDefaultSharedPreferences(App.getAppContext());
        if (defaultSharedPreferences.getBoolean("notifications", true))
            FirebaseMessaging.subscribeToTopics();
        else
            FirebaseMessaging.unsubscribeFromTopics();
    }

    private void setupViewPager(ViewPager viewPager) {
        ViewPagerAdapter adapter = new ViewPagerAdapter(getSupportFragmentManager());

        Fragment postsFragment = new PostsFragment();
        Bundle bundle1 = new Bundle();
        bundle1.putInt("layout", R.layout.post_with_avatars);
        postsFragment.setArguments(bundle1);
        adapter.addFragment(postsFragment, getResources().getString(R.string.posts));

        Fragment eventsFragment = new EventsFragment();
        Bundle bundle2 = new Bundle();
        bundle2.putInt("layout", R.layout.row_event_with_avatars);
        eventsFragment.setArguments(bundle2);
        adapter.addFragment(eventsFragment, getResources().getString(R.string.events));

        adapter.addFragment(new ClubsFragment(), getResources().getString(R.string.clubs));
        adapter.addFragment(new NotificationsFragment(), getResources().getString(R.string.notifications));

        viewPager.setAdapter(adapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);

        MenuItem menuItem = menu.findItem(R.id.search);

        SearchView searchView = (SearchView) MenuItemCompat.getActionView(menuItem);
        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        searchView.setSearchableInfo(searchManager.getSearchableInfo(new ComponentName(this, SearchActivity.class)));

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_profile:
                startActivity(new Intent(this, ProfileActivity.class));
                break;

            case R.id.action_settings:
                startActivity(new Intent(this, SettingsActivity.class));
                break;

            case R.id.action_legal_conditions:
                startActivity(new Intent(this, LegalConditionsActivity.class));
                break;

            case R.id.action_credits:
                startActivity(new Intent(this, CreditsActivity.class));
                break;

            default:
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (MainActivity.customTabsConnection != null) {
            this.unbindService(MainActivity.customTabsConnection);
            MainActivity.customTabsConnection = null;
        }
    }

    private void ifHuaweiAlert() {
        if (Build.MANUFACTURER.equalsIgnoreCase("huawei")) {
            final SharedPreferences defaultSharedPreferences = PreferenceManager.getDefaultSharedPreferences(App.getAppContext());

            if (!defaultSharedPreferences.getBoolean("protected_apps", false)) {
                final SharedPreferences.Editor editor = defaultSharedPreferences.edit();

                Intent intent = new Intent();
                intent.setClassName("com.huawei.systemmanager", "com.huawei.systemmanager.optimize.process.ProtectActivity");

                if (isCallable(intent)) {
                    final AppCompatCheckBox dontShowAgain = new AppCompatCheckBox(this);
                    dontShowAgain.setText(getString(R.string.protected_apps_skip));
                    dontShowAgain.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                        @Override
                        public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                            editor.putBoolean("protected_apps", isChecked);
                            editor.apply();
                        }
                    });

                    new AlertDialog.Builder(this)
                            .setTitle(getString(R.string.protected_apps_dialog_title))
                            .setMessage(String.format(getString(R.string.protected_apps_dialog_message), getString(R.string.app_name)))
                            .setView(dontShowAgain)
                            .setPositiveButton(getString(R.string.protected_apps_button), new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog, int which) {
                                    huaweiProtectedApps();
                                }
                            })
                            .setNegativeButton(R.string.close_button, null)
                            .show();
                }
                else {
                    editor.putBoolean("protected_apps", true);
                    editor.apply();
                }
            }
        }
    }

    private boolean isCallable(Intent intent) {
        final List<ResolveInfo> list = getPackageManager().queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);

        return list.size() > 0;
    }

    private void huaweiProtectedApps() {
        try {
            Runtime.getRuntime().exec("am start -n com.huawei.systemmanager/.optimize.process.ProtectActivity --user " + getUserSerial());
        }
        catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    private String getUserSerial() {
        //noinspection ResourceType
        Object userManager = getSystemService("user");

        if (null == userManager) {
            return "";
        }

        try {
            Method myUserHandleMethod = android.os.Process.class.getMethod("myUserHandle", (Class<?>[]) null);
            Object myUserHandle = myUserHandleMethod.invoke(android.os.Process.class, (Object[]) null);
            Method getSerialNumberForUser = userManager.getClass().getMethod("getSerialNumberForUser", myUserHandle.getClass());
            Long userSerial = (Long) getSerialNumberForUser.invoke(userManager, myUserHandle);

            if (userSerial != null) {
                return String.valueOf(userSerial);
            }
            else {
                return "";
            }
        } catch (NoSuchMethodException | IllegalArgumentException | InvocationTargetException | IllegalAccessException ex) {
            ex.printStackTrace();
        }

        return "";
    }
}

File: app/src/main/java/fr/insapp/insapp/activities/MainActivity.kt
package fr.insapp.insapp.activities

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.SearchManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatCheckBox
import androidx.appcompat.widget.SearchView
import androidx.browser.customtabs.CustomTabsClient
import androidx.preference.PreferenceManager
import androidx.viewpager.widget.ViewPager
import com.google.android.gms.tasks.OnCompleteListener
import com.google.firebase.iid.FirebaseInstanceId
import fr.insapp.insapp.App
import fr.insapp.insapp.R
import fr.insapp.insapp.adapters.ViewPagerAdapter
import fr.insapp.insapp.components.CustomTabsConnection
import fr.insapp.insapp.fragments.ClubsFragment
import fr.insapp.insapp.fragments.EventsFragment
import fr.insapp.insapp.fragments.NotificationsFragment
import fr.insapp.insapp.fragments.PostsFragment
import fr.insapp.insapp.notifications.MyFirebaseMessagingService
import fr.insapp.insapp.utility.Utils
import kotlinx.android.synthetic.main.activity_main.*
import java.io.IOException

class MainActivity : AppCompatActivity() {

    companion object {
        @JvmField var customTabsConnection: CustomTabsConnection? = null
    }

    private val userSerial: String
        get() {
            val userManager = getSystemService(Context.USER_SERVICE) ?: return ""

            try {
                val myUserHandleMethod = Process::class.java.getMethod("myUserHandle", null)
                val myUserHandle = myUserHandleMethod.invoke(Process::class.java, null)
                val getSerialNumberForUser = userManager.javaClass.getMethod("getSerialNumberForUser", myUserHandle.javaClass)
                val userSerial = getSerialNumberForUser.invoke(userManager, myUserHandle) as Long

                return userSerial.toString()
            } catch (ex: Exception) {
                ex.printStackTrace()
            }

            return ""
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // toolbar
        setSupportActionBar(toolbar_main)
        supportActionBar?.setDisplayHomeAsUpEnabled(false)
        supportActionBar?.title = Utils.user?.username

        // view pager
        setupViewPager(viewpager)

        // tab layout
        tabs.setupWithViewPager(viewpager)

        // custom tabs optimization
        customTabsConnection = CustomTabsConnection()
        CustomTabsClient.bindCustomTabsService(this, "com.android.chrome", customTabsConnection)

        // display correct fragment for shortcuts
        val fragmentId = intent.getIntExtra("FRAGMENT_ID", 0)
        viewpager.currentItem = fragmentId

        // Huawei protected apps
        ifHuaweiAlert()

        // notification channels (android O and above)
        createNotificationChannel("posts", getString(R.string.notification_posts_name), getString(R.string.notification_posts_description))
        createNotificationChannel("events", getString(R.string.notification_events_name), getString(R.string.notification_events_description))

        // Firebase token
        FirebaseInstanceId.getInstance().instanceId
            .addOnCompleteListener(OnCompleteListener { task ->
                if (!task.isSuccessful) {
                    Log.w(MyFirebaseMessagingService.TAG, "getInstanceId failed", task.exception)
                    return@OnCompleteListener
                }

                // Get new Instance ID token
                val token = task.result?.token
                Log.d(MyFirebaseMessagingService.TAG, "Current Firebase token: $token")
            })

        updateApp()
    }

    //TODO: delete after update
    private fun updateApp() {
        val defaultSharedPreferences = PreferenceManager.getDefaultSharedPreferences(App.getAppContext())

        val oldSex = defaultSharedPreferences.getString("sex", "")
        if (!oldSex.isNullOrEmpty()) {
            defaultSharedPreferences.edit().putString("gender", oldSex).apply()
            defaultSharedPreferences.edit().remove("sex").apply()
        }

        val oldClass = defaultSharedPreferences.getString("class", "")
        if (!oldClass.isNullOrEmpty() && oldClass == "Alternant") {
            defaultSharedPreferences.edit().putString("class", "5CDTI").apply()
        }

        MyFirebaseMessagingService.subscribeToTopic("news", false)
        MyFirebaseMessagingService.subscribeToTopic("events", false)
    }

    private fun setupViewPager(viewPager: ViewPager) {
        val adapter = ViewPagerAdapter(supportFragmentManager)

        val postsFragment = PostsFragment()
        val bundle1 = Bundle()
        bundle1.putInt("layout", R.layout.post_with_avatars)
        postsFragment.arguments = bundle1
        adapter.addFragment(postsFragment, resources.getString(R.string.posts))

        val eventsFragment = EventsFragment()
        val bundle2 = Bundle()
        bundle2.putInt("layout", R.layout.row_event_with_avatars)
        eventsFragment.arguments = bundle2
        adapter.addFragment(eventsFragment, resources.getString(R.string.events))

        adapter.addFragment(ClubsFragment(), resources.getString(R.string.clubs))
        adapter.addFragment(NotificationsFragment(), resources.getString(R.string.notifications))

        viewPager.adapter = adapter
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)

        val searchItem = menu.findItem(R.id.search)
        val searchView = searchItem.actionView as SearchView
        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
        searchView.setSearchableInfo(searchManager.getSearchableInfo(ComponentName(this, SearchActivity::class.java)))

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_profile -> startActivity(Intent(this, ProfileActivity::class.java))

            R.id.action_settings -> startActivity(Intent(this, SettingsActivity::class.java))

            R.id.action_how_to_post -> startActivity(Intent(this, HowToPostActivity::class.java))

            R.id.action_legal_conditions -> startActivity(Intent(this, LegalConditionsActivity::class.java))

            R.id.action_credits -> startActivity(Intent(this, CreditsActivity::class.java))
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()

        if (customTabsConnection != null) {
            this.unbindService(customTabsConnection)
            customTabsConnection = null
        }
    }

    private fun ifHuaweiAlert() {
        if (Build.MANUFACTURER.equals("huawei", ignoreCase = true)) {
            val defaultSharedPreferences = PreferenceManager.getDefaultSharedPreferences(App.getAppContext())

            if (!defaultSharedPreferences.getBoolean("protected_apps", false)) {
                val editor = defaultSharedPreferences.edit()

                val intent = Intent().setClassName("com.huawei.systemmanager", "com.huawei.systemmanager.optimize.process.ProtectActivity")

                if (isCallable(intent)) {
                    val dontShowAgain = AppCompatCheckBox(this)
                    dontShowAgain.text = getString(R.string.protected_apps_skip)
                    dontShowAgain.setOnCheckedChangeListener { _, isChecked ->
                        editor.putBoolean("protected_apps", isChecked)
                        editor.apply()
                    }

                    AlertDialog.Builder(this, R.style.InsappMaterialTheme_AlertDialogTheme)
                            .setTitle(getString(R.string.protected_apps_dialog_title))
                            .setMessage(String.format(getString(R.string.protected_apps_dialog_message), getString(R.string.app_name)))
                            .setView(dontShowAgain)
                            .setPositiveButton(getString(R.string.protected_apps_button)) { _, _ -> huaweiProtectedApps() }
                            .setNegativeButton(R.string.close_button, null)
                            .show()
                } else {
                    editor.putBoolean("protected_apps", true)
                    editor.apply()
                }
            }
        }
    }

    private fun isCallable(intent: Intent): Boolean {
        return packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY).size > 0
    }

    private fun huaweiProtectedApps() {
        try {
            Runtime.getRuntime().exec("am start -n com.huawei.systemmanager/.optimize.process.ProtectActivity --user $userSerial")
        } catch (ex: IOException) {
            ex.printStackTrace()
        }
    }


    private fun createNotificationChannel(id: String, name: String, description: String) {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(id, name, NotificationManager.IMPORTANCE_DEFAULT)
            channel.description = description
            // Register the channel with the system; you can't change the importance
            // or other notification behaviors after this
            val notificationManager = getSystemService(NotificationManager::class.java)
            notificationManager?.createNotificationChannel(channel)
        }
    }
}
